src/java_bytecode/java_static_initializers.cpp

```diff
@@ -96,21 +204,41 @@ static void create_clinit_wrapper_symbols(
   symbol_tablet &symbol_table,
   synthetic_methods_mapt &synthetic_methods)
 {
-  const irep_idt &already_run_name =
-    clinit_already_run_variable_name(class_name);
-  symbolt already_run_symbol;
-  already_run_symbol.name = already_run_name;
-  already_run_symbol.pretty_name = already_run_name;
-  already_run_symbol.base_name = "clinit_already_run";
-  already_run_symbol.type = bool_typet();
-  already_run_symbol.value = false_exprt();
-  already_run_symbol.is_lvalue = true;
-  already_run_symbol.is_state_var = true;
-  already_run_symbol.is_static_lifetime = true;
-  already_run_symbol.mode = ID_java;
-  bool failed = symbol_table.add(already_run_symbol);
-  INVARIANT(!failed, "clinit-already-run symbol should be fresh");
-
+  exprt zero =
+    from_integer(
+      static_cast<int>(clinit_statest::NOT_INIT), clinit_states_type);
+
+  // Create two global static synthetic "fields" for the class "id"
+  // these two variables hold the state of the class initialization algorithm
+  // across calls to the clinit_wrapper
+  add_new_symbol(
+    symbol_table,
+    clinit_state_var_name(class_name),
+    clinit_states_type,
+    zero,
+    false,
+    true);
+
+  add_new_symbol(
+    symbol_table,
+    clinit_thread_local_state_var_name(class_name),
+    clinit_states_type,
+    zero,
+    true,
+    true);
+
+  // Create a local variable "init_complete". This variable is used as an
```

Comment:
I think this comment is inaccurate, does it belong to the thread-local variable instead? The function-local one is just used to copy the global so you can inspect it and then maybe return outside a critical section.

Reply:
the comment is misleading agreed. rewording.

Reply:
We are using a local boolean for inspection on purpose. The reason being that inspecting the global state  resulted in symex related propagation issues in multi-threaded code.